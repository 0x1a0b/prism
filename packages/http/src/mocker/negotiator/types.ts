import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}
