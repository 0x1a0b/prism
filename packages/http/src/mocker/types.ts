import { IOperationOptions, Protocol } from '@stoplight/prism-core/mocker/types';

export interface IHttpOperationOptions extends IOperationOptions {
  protocol: Protocol.HTTP;
  // Defaults to the lowest 2xx status code
  status: string;
  // Defaults to application/json
  mediaType: string;
  // Defaults to content type
  example?: string;
  // If true mock back response headers
  headers?: boolean;
}
