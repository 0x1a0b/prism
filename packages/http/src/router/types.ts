import { IHttpOperation } from '@stoplight/types/http';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperation;
  serverMatch: MatchType | null;
  pathMatch: MatchType;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
