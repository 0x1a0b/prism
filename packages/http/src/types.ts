import {
  IFilesystemLoaderOpts,
  IPrism,
  IPrismComponents,
  IPrismConfig,
} from '@stoplight/prism-core';
import { IHttpOperation } from '@stoplight/types';

export type TPrismHttpInstance<LoaderInput = IFilesystemLoaderOpts> = IPrism<
  IHttpOperation,
  IHttpRequest,
  IHttpResponse,
  IHttpConfig,
  LoaderInput
>;

export type TPrismHttpComponents<LoaderInput = IFilesystemLoaderOpts> = Partial<
  IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, LoaderInput>
>;

// TODO: should be complete | and in the @stoplight/types repo
export type IHttpMethod =
  | 'get'
  | 'put'
  | 'post'
  | 'delete'
  | 'options'
  | 'head'
  | 'patch'
  | 'trace'; // ... etc

export interface IHttpOperationConfig {
  mediaType?: string;
  code?: string;
  exampleKey?: string;
  dynamic?: boolean;
}

export interface IHttpConfig extends IPrismConfig {
  mock: boolean | IHttpOperationConfig;

  security?: {
    // TODO
  };

  validate?: {
    request?:
      | boolean
      | {
          hijack?: boolean;
          headers?: boolean;
          query?: boolean;
          body?: boolean;
        };

    response?:
      | boolean
      | {
          headers?: boolean;
          body?: boolean;
        };
  };
}

export interface IHttpRequest {
  method: IHttpMethod;
  url: {
    baseUrl?: string;
    path: string;
    query?: {
      [name: string]: string;
    };
  };
  headers?: {
    [name: string]: string;
  };
  body?: any;
}

export interface IHttpResponse {
  statusCode: number;
  headers?: {
    [name: string]: string;
  };
  body?: any;
}
