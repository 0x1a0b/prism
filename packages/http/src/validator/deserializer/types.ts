import { ISchema } from '@stoplight/types/schema';

export type DeserializeHttpHeader = (value: string, type: string, explode: boolean) => any;

export interface IHttpHeaderParamStyleDeserializer {
  supports: (style: string) => boolean;
  deserialize: DeserializeHttpHeader;
}

export interface IHttpParamDeserializerRegistry<T extends Function> {
  get(style: string): T | undefined;
}

export interface IHttpParamStyleDeserializer<T> {
  supports: (style: string) => boolean;
  deserialize: T;
}

export type DeserializeHttpQuery = (
  key: string,
  query: {
    [name: string]: string | string[];
  },
  schema: ISchema,
  explode: boolean
) => any;

export interface IHttpQueryParamStyleDeserializer
  extends IHttpParamStyleDeserializer<DeserializeHttpQuery> {
  deserialize: DeserializeHttpQuery;
}
