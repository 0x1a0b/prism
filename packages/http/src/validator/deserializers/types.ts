import { HttpParamStyles, ISchema } from '@stoplight/types';

import { IHttpNameValue, IHttpNameValues } from '../../types';

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: (name: string, parameters: Parameters, schema: ISchema, explode?: boolean) => any;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
