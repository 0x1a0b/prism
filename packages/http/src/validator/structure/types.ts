import { IValidation } from '@stoplight/prism-core';

export interface IHttpValidator<Obj, Specs> {
  validate(obj: Obj, specs: Specs[], mediaType?: string): IValidation[];
}
