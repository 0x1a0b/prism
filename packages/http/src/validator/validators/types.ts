import { IValidation } from '@stoplight/prism-core';
import { ISchema } from '@stoplight/types/schema';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IValidation[];
}

export interface ISchemaValidator<S extends ISchema> {
  validate(content: any, schema: S): IValidation[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: ISchema) => IValidation[]) | undefined;
}
