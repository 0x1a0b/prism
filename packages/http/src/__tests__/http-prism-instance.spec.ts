import { IPrism } from '@stoplight/prism-core';
import { IHttpOperation } from '@stoplight/types/http-spec';
import { omit } from 'lodash';
import { resolve } from 'path';
import { createInstance, IHttpConfig, IHttpRequest, IHttpResponse } from '../';

describe('Http Prism Instance function tests', () => {
  let prism: IPrism<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, { path: string }>;

  beforeAll(async () => {
    prism = createInstance();
    await prism.load({
      path: resolve(__dirname, 'fixtures', 'no-refs-petstore-minimal.oas2.json'),
    });
  });

  test('given incorrect route should throw error', () => {
    return expect(
      prism.process({
        method: 'get',
        url: {
          path: '/invalid-route',
        },
      })
    ).rejects.toThrowError('Route not resolved, none path matched');
  });

  test('given correct route should return correct response', async () => {
    const response = await prism.process({
      method: 'get',
      url: {
        path: '/pet/findByStatus',
        query: {
          status: ['available', 'pending'],
        },
      },
    });
    const parsedBody = JSON.parse(response!.output!.body);
    expect(parsedBody.length).toBeGreaterThan(0);
    parsedBody.forEach((element: any) => {
      expect(typeof element.name).toEqual('string');
      expect(Array.isArray(element.photoUrls)).toBeTruthy();
      expect(element.photoUrls.length).toBeGreaterThan(0);
    });
    // because body is generated randomly
    expect(omit(response, 'output.body')).toMatchSnapshot();
  });

  test('given route with invalid param should return a validation error', async () => {
    const response = await prism.process({
      method: 'get',
      url: {
        path: '/pet/findByStatus',
      },
    });
    expect(response).toMatchSnapshot();
  });
});
