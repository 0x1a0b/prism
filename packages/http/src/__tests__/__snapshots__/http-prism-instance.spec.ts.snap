// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Http Prism Instance function tests given correct route should return correct response 1`] = `
Object {
  "input": Object {
    "method": "get",
    "url": Object {
      "path": "/pet/findByStatus",
      "query": Object {
        "status": Array [
          "available",
          "pending",
        ],
      },
    },
  },
  "output": Object {
    "headers": Object {
      "Content-type": "application/json",
    },
    "statusCode": 200,
  },
  "validations": Object {
    "input": Array [],
    "output": Array [],
  },
}
`;

exports[`Http Prism Instance function tests given route with invalid param should return a validation error 1`] = `
Object {
  "input": Object {
    "method": "get",
    "url": Object {
      "path": "/pet/findByStatus",
    },
  },
  "output": Object {
    "body": "ERROR: Your request is not valid.
We cannot generate a sensible response because your '400'
response has neither example nor schema or is not defined.
Here is the original validation result instead: [{\\"path\\":[\\"query\\",\\"status\\"],\\"name\\":\\"required\\",\\"summary\\":\\"\\",\\"message\\":\\"Missing status query param\\",\\"severity\\":\\"error\\"}]",
    "headers": Object {
      "Content-type": "text/plain",
    },
    "statusCode": 400,
  },
  "validations": Object {
    "input": Array [
      Object {
        "message": "Missing status query param",
        "name": "required",
        "path": Array [
          "query",
          "status",
        ],
        "severity": "error",
        "summary": "",
      },
    ],
    "output": Array [],
  },
}
`;
